### PR TITLE
Adding stat_summary to list of geoms 

### LINKED ISSUE #4420: Adding stat_summary as a "geom"
There's a lot of `stat_???` functions in `ggplot2` which we will eventually want to implement. However, I think `stat_summary` is extremely useful even if this is the only one we have to start. 

It is coming up a lot with the agriculture type data for example, adding means or medians as points/lines on top of a plot. This is possible through our general graphics by using layers with different data frames but stat_summary does this very easily.

I believe the structure fits in to our structure of geoms in that it has aesthetics and other parameters. So I think it can just be added like a geom. @maxwellfundi can you investigate this by trying to add this as a "geom" in R-Instat? We might just want to add another property to our structure to say whether something is a "geom" or a "stat".

### PR DESCRIPTION
**Pull request by maxwellfundi**
@africanmathsinitiative/developers this could now be tested.

Fixes #4420

### CODE REVIEW

**Comment by Muthenya** [comment]
This works fine from my end. @dannyparsons this is ready for you.

File: instat/ucrGeom.vb [hunk 1]
```diff
@@ -159,6 +159,39 @@ Public Class ucrGeom
...
+        'mandatory aesthetics
+        clsgeom_statsummary.AddAesParameter("x", strIncludedDataTypes:={"factor", "numeric"}, bIsMandatory:=True)
+        clsgeom_statsummary.AddAesParameter("y", strIncludedDataTypes:={"factor", "numeric"}, bIsMandatory:=True)
+        clsgeom_statsummary.AddAesParameter("group", strIncludedDataTypes:={"factor", "numeric"})
+
+        'layer params
+
+        'These are the list of parameters on the documentation, some may have more options than what we have here which am yet to find out
+        clsgeom_statsummary.AddLayerParameter("fun.data", "list", Chr(34) & "mean_cl_boot" & Chr(34), lstParameterStrings:={Chr(34) & "mean_sdl" & Chr(34), Chr(34) & "mean_cl_boot" & Chr(34)})
+        clsgeom_statsummary.AddLayerParameter("fun.y", "list", Chr(34) & "mean" & Chr(34), lstParameterStrings:={Chr(34) & "median" & Chr(34), Chr(34) & "mean" & Chr(34)})
```
[the remark this answers is not in the document]
Comment (dannyparsons) [reply instat/ucrGeom.vb]: Yes some common summaries in the combo box but editable as well. I think this means we now want two types of "list" parameters, editable and non-editable, which could be done in the specification of the parameter. Can you look into doing this?

[the remark this answers is not in the document]
Comment (maxwellfundi) [reply instat/ucrGeom.vb]: Yeah I can do that :-)

**Comment by bethanclarke** [comment]
Hi @maxwellfundi, so I thought I'd test this by trying to create the following graph:
statsummary

Unfortunately, I R-Instat produced a graph that just looked like a scatter plot. When I ran this code in R I got the following warning: 
Computation failed in `stat_summary()`: Hmisc package required for this function 

After installing Hmisc, I got the graph above. The only thing is, I can't find where Hmisc is being used, so the `Hmisc::` notation wouldn't work here.

**Comment by dannyparsons** [comment]
It's being used inside "mean_cl_boot" so yes we won't be able to add `Hmisc::` anywhere to fix this. Once Hmisc is installed it should then work. So if we include Hmisc as a package in R-Instat this should work.

**Comment by bethanclarke** [comment]
stat_summary(fun.y = "mean", colour = "red", size =2) and stat_summary(fun.y = "median", colour = "red", size =2) do not work

For this to work for the `fun.y ` option, we also need to include `geom = "point"` into `stat_summary ` i.e.
stat_summary(fun.y = "mean", colour = "red", size =2, geom= "point") and the same for median option

**Comment by bethanclarke** [comment]
@maxwellfundi This is where I am getting the graphs to test: 
http://ggplot2.tidyverse.org/reference/stat_summary.html

**Comment by maxwellfundi** [comment]
@bethanclarke thanks for testing this work. I used that documentation you were testing with to implement this geom.  We currently have no good restrictions and linking of dependent parameters(was started but not working now) and so i think people need to be more careful here. 

@dannyparsons We had the issue of `bIsDependent` can we get this to work properly to avoid these kind of issues that @bethanclarke  has raised?

**Comment by dannyparsons** [comment]
Do we have mandatory layer parameters yet?

**Comment by dannyparsons** [comment]
Hmisc package now added which solve the issue raised here https://github.com/africanmathsinitiative/R-Instat/pull/4600#issuecomment-391290946

**Comment by dannyparsons** [comment]
Has any progress been made on this? I would quite like to test this out for the next version.

**Comment by maxwellfundi** [comment]
@dannyparsons  I am now working on this. I have added. 

So i have added this an `editablelist` as the other type of list we can have for the layer parameters then added the line to make the combo box to be editable on the `ucrReceiverMetadataProperty`
```

 ElseIf clsLayerParam.strLayerParameterDataType = "editablelist" Then
                ctrActive = ucrInputCboParamValue
                ucrInputCboParamValue.SetDropDownStyleAsEditable(True)
```
I then looked at the method on the ucrinputcombo and saw there was something that needed to be implemented here. The validation 

   ```
 Public Sub SetDropDownStyleAsEditable(bAdditionsAllowed As Boolean)
        cboInput.DropDownStyle = ComboBoxStyle.DropDown
        cboInput.AutoCompleteMode = AutoCompleteMode.Append
        cboInput.AutoCompleteSource = AutoCompleteSource.ListItems
        'TODO implement validation settings for this		
        If bAdditionsAllowed Then

        Else

        End If
    End Sub
```

So I am not sure what sort of validation we need here because any of the things we need could be anything really. So what validations did you have in mind for this?

**Comment by maxwellfundi** [comment]
In addittion, following to your question above we only have mandatory aesthetics and that actually works.

**Comment by dannyparsons** [comment]
That looks good. But then the combo box needs to be made non editable again in the other case.

The validation part is to adjust the validation settings in these cases, which is still to be done. But it doesn't effect the combo boxes working correctly now. It's an enhancement that is yet to be done.

**Comment by maxwellfundi** [comment]
Sounds good @dannyparsons. I will go ahead and change the other cases to be noneditable.

**Comment by maxwellfundi** [comment]
@bethanclarke  I have found out that because of `aes(colour = factor(vs))` on the fourth graph you could not be able create it. 

You needed to have the third receiver filled in the scatterplot dialog(we also need to add the factor()) so that the can have that graph. 

Alternatively, @bethanclarke  you needed to convert vs variable to factor and you would have exactly the same graph.  So this works.

**Comment by maxwellfundi** [comment]
@bethanclarke  you can now retest this

@africanmathsinitiative/developers This is ready for review

**Comment by bethanclarke** [comment]
Okay I will test this today :)

**Comment by maxwellfundi** [comment]
@bethanclarke ?

**Comment by bethanclarke** [comment]
I have tested this and it is working well, @dannyparsons this is ready for testing, thanks.

**Pull request merged**